netket/vqs/mc_state.py

```diff
@@ -361,6 +374,56 @@ def n_discard(self, val) -> int:
         )
         self.n_discard_per_chain = val
 
+    @property
+    def minibatch_size(self) -> int:
+        return self._minibatch_size
+
+    @minibatch_size.setter
+    def minibatch_size(self, minibatch_size: int):
+        # disable minibatches if it is None
+        if minibatch_size is None:
+            self._minibatch_size = None
+            return
+
+        do_warn = False
+
+        # must be a multiple of n_chains
+        if minibatch_size % self.sampler.n_chains_per_rank != 0:
```

Comment:
Sorry to repeat myself, but I feel very strongly that we should get rid of this requirement. Why does `minibatch_size/n_chains` have to be an integer? 

This is quite prohibitive because a `expect_and_grad` takes `minibatch_size x neighbors x model_memory` memory whereas sampling only takes `n_chains x model_memory`.  So there's a ton of situations where you want to set n_chains > minibatch_size

Reply:
I was on holiday and read your comments but not answered...
I am aware of this very valid point.

My main issue with batching this separately is that it makes the interface uglier for arbitrary operators, but that's entirely possible.

Another issue is that if all mini batches have the same size, we don't need to compile the code twice.
Otherwise we do, and we need to make a more complex implementation.

I will get back to this and probably change the implementation.
Also it might make the implementation nicer to make the operators Jax-compatible. 
I'd like to find a way to make the implementation nice.
I'll get back to this in a month or two.

Reply:
I realize I'm just raising the same objection as everyone else... but I do think it's a very important one

Reply:
> Another issue is that if all mini batches have the same size, we don't need to compile the code twice.
> Otherwise we do, and we need to make a more complex implementation.

I think it's entirely reasonable to require that n_minibatch divide the total number of samples (per MPI process), and error out if you supply one that doesn't rather than trying to match the sample number to it, which would be a nightmare if you also match to `n_chain` (not sure if that's super useful either)

> My main issue with batching this separately is that it makes the interface uglier for arbitrary operators, but that's entirely possible.

Why is that? Operators would still get batches of shape `[minibatch_size, system_size]`, do they care about how they were sampled?

Reply:
The complication arises in the kernels: https://github.com/netket/netket/blob/69b009813d233d472f33f8f762dc58b79235c8e6/netket/vqs/mc_expect.py#L52

Right now the interface of the `operator kernels` is that they take as input `(logpsi, pars, σ, σp, mel)`.
This works well with (weird) kernels like the one for operators*densitymatrix that manipulates `sigmap` and `sigma`.

The kernel is implemented for one scalar `sigma` (and a vector sigmap).
The it is vmapped.

To properly batch along `sigmap`, we'd have to 1) require that those kernels are written for the whole batch of sigma and sigmap, and eventually for them to implement minibatching also (this might lead to code duplciation?) or 2) pre-compute `psi(sigma)` and `psi(sigmap)` and feed it later to the kernel, but that might limit the type of kernels we can write, and wouldnt play well with `operator*densitymtarix`.

Reply:
I'm not sure what this has to do with minibatching. All you need to do is splitting `σ` and `σp` into chunks, feed it into this function in a loop, and sum the outputs. All the kernel would notice is that it's called with fewer inputs

Reply:
The kernel must be called with N sigma and N*M sigmap. 
So your effective batch size is a multiple of M, not whatever the user selected.

And what if you want minibatches of size smaller than M?
You cant.
Maybe that's asking too much?

Reply:
M is the number of neighbours in Hilbert space right? That should be fine, in practice we're talking about reducing the N you pull through in one cycle to the order of 10-100, without restricting the total sample count

Reply:
yes. Or number of nonzero entries per row of an operator, as i'd put it.

Reply:
> I think literally all we want is decoupling from the number of MC chains, which has nothing whatsoever to do with calculating expectation values (see #830 (comment))

I understand. but that's not as simple as saying it while retaining a decent interface for extending it.

Reply:
If N is the number of samples and M is the maximum number of neighbors, all you need to so is have the samples be shape [N_mini, L_mini, H_size] and the neighbors be split [N_mini, L_mini, M, H_size] then loop over the first index and vmap over the second. 

RIght now N has to be a multiple of N_{chains} which is unnecessary.

Reply:
The problem of the current implementation is that it is batching the passes, using the same minibatch size for the forward and backward pass alltogether (which is why the implementation is simpler). It's also why it constraints the n_samples to be a multiple of n_chains and minibatch size. (It could not enforce this constraint, but then you'd double compile times, that's why the constraint is in place). 
That had the added benefit that i could literally just use the old code, plus some boilerplate.

using (mini)batches only when evaluating `logpsi`, which is what everyone keeps proposing (because it makes sense, as forward and backward passes shouldn't be batched alltogether)  is more involved, because one must change the whole logic.

Reply:
@chrisrothUT as i wrote above, that requirement can be relaxed, it's there only to ensure that compilation times don't double.

Reply:
> n_samples to be a multiple of n_chains and minibatch_size

This is completely fine, what I'm arguing is that minibatch_size doesn't have to be a multiple of n_chains

Reply:
In any case, I don't think avoiding a jit-compilation is worth that kind of strong restriction. At least in my situation, I'd have to create a whole new branch just to avoid that restriction.

Reply:
out of curiosity, what would be the minibatch size and number of chains you use?

Reply:
In most of my work, I do 1024 samples, 1024 chains, and minibatches of 64/128 depending on the net. (also n_discard is set to 0 or 1 because mixing is usually good enough)

Reply:
but 64 is a minibatch of the forward pass, so the actual size of sigmap in a minibatch will be `l_minibatch x M`, right?

Reply:
in case you're interested, this is the bootleg version Chris and I use, it's not supporting every fine detail of statistics yet, but it produces correct expectation values https://github.com/attila-i-szabo/netket/blob/5c94b76a16b607b13f20f78392a473524344ec33/netket/vqs/mc_expect_grad_minibatch.py#L43

Reply:
the fact that M can be very large (even much larger than the number of sites, in some cases) is problematic and should be taken into account in the mini batch logic, otherwise we will end up implementing something that will still break the memory for some hamiltonians

Reply:
> using (mini)batches only when evaluating `logpsi`, which is what everyone keeps proposing (because it makes sense, as forward and backward passes shouldn't be batched alltogether) is more involved, because one must change the whole logic.

this is also dealt with in the above version, all you do is flatten `σ` back to 2D before calling the backward pass

Reply:
I don't know how easy that is, but at least to compute the local energy it should be possible to first collect all sigma and sigma_p into a single array and then mini-batch that with the guarantee that the forward pass is never called for more than some number of samples

Reply:
> I don't know how easy that is, but at least to compute the local energy it should be possible to first collect all sigma and sigma_p into a single array and then mini-batch that with the guarantee that the forward pass is never called for more than some number of samples

At this point it's probably better to minibatch `logpsi` itself in some way that if you try to pull through more than some number of inputs, it breaks it down (although that would probably lead to a lot of rejitting unless you choose the cutoff wisely)

Reply:
> At this point it's probably better to minibatch logpsi itself in some way that if you try to pull through more than some number of inputs, it breaks it down (although that would probably lead to a lot of rejitting unless you choose the cutoff wisely)

yes, exactly, if that could be implemented efficiently it would be just a simpler solution (maybe just adding a method that calls the forward pass looping over the extra dimension). Of course the issue is wether this is efficient or not...

Reply:
> At this point it's probably better to minibatch logpsi itself in some way that if you try to pull through more than some number of inputs, it breaks it down (although that would probably lead to a lot of rejitting unless you choose the cutoff wisely)

That's not really doable, because we use `vmap` in so many places.
so you would be vmapping over a loop which makes no sense.

Reply:
no wait, I am saying to loop over a vmap, not the contrary. you add a method _minibatch_forward or something, that internally calls the main forward method (defined with vmap or anything)

Reply:
also it doesn't solve the issue that VJPs use more memory than forward passes, so you'd want to use different cutoffs for the two operations.

Reply:
> also it doesn't solve the issue that VJPs use more memory than forward passes, so you'd want to use different cutoffs for the two operations.

it should be a constant multiplicative value around at most 3 for most cases, i think.
It's a lot of effort to have different values for a relatively small chagne,.

Reply:
something very inefficient might be 

```python
def _mini_batch_forward(x):
    for i in range(x.shape[0]//n_mini):
           ... self.forward(x[i:i+mini_size])
  
   return out
```

and you always call this _mini_batch_forward instead of forward in all the functions you have in expect_grad

Reply:
> no wait, I am saying to loop over a vmap, not the contrary. you add a method _minibatch_forward or something, that internally calls the main forward method (defined with vmap or anything)

assuming forward means the whole `expect` function, this is what this pr does as of now.
assuming forward is the model itself, this is not as simple and requries a rewrite of most of those methods, because we assume we can call `vmap` everywhere.  I'm not saying it's impossible. I'm just saying it's ugly and makes it harder to add new operator kernels.

Reply:
> it should be a constant multiplicative value around at most 3 for most cases, i think.

that's runtime. memory is worse because in a deep network you have to store gradients for every nonlinear layer, which you don't have with the forward pass (you just overwrite values in previous layers). so it's at least a factor of (network depth) and can get even worse if you have more complex operations than just feed-forward

Reply:
although in `expect_and_grad` as it's minibatched now, you call forward passes with much bigger inputs, so that should offset it. in any case, you might end up with situations where you have to minibatch QGTOnTheFly...

Reply:
> assuming forward is the model itself, this is not as simple and requries a rewrite of most of those methods, because we assume we can call vmap everywhere. I'm not saying it's impossible. I'm just saying it's ugly and makes it harder to add new operator kernels. Probably solvable in some way, but would take some experimentation to get it right.

I understand, but maybe this is telling us that we actually don't need to assume/use vmap on such (very) large batches, that instead should just be handled with a for without loss of efficiency? Maybe the solution is just removing vmaps in the kernels...

Reply:
my impression is that if one uses vmaps on batches of size of thousands or more there is no intrinsic benefit and it might actually even hurt performance (maybe not on 20 spins, but that's a marginal case we can sacrifice)

Reply:
> in any case, you might end up with situations where you have to minibatch QGTOnTheFly...

I've just produced a case where `expect_and_grad_minibatch` runs out of memory on the backward pass, but QGTOnTheFly (which has to do the same VJP) doesn't. I was trying to look but I've no clue what causes this, but it's certainly not great

Reply:
I think if we got rid of vmap entirely, there would be a pretty big performance hit on small systems. It's true that you don't gain much when you're near max memory but you'll be pretty far from that looping over one sample at a time.

I like our current implementation, which takes max memory `L_mini x M x H_size`. Obviously M isn't constant and you can theoretically get burned very late in the calc with an anomalous sample (happened to me once) but usually you'll get something close to the maximum M over the whole calculation in the first few iterations. We could also give a warning if you're over 80% memory or something.

Reply:
>  think if we got rid of vmap entirely, there would be a pretty big performance hit on small systems. It's true that you don't gain much when you're near max memory but you'll be pretty far from that looping over one sample at a time.

I am not saying you should loop over one sample at the time, I was thinking about still vmapping but only on the mini batch (on 20 spins you can still max that...)

Ideally, what we really want (I think) is a vmap that accepts an extra argument (mini_batch_size) and that starts looping instead of vectorizing, if the input size exceeds mini_batch_size. The problem is that now we are using vmap everywhere...

Reply:
> Ideally, what we really want (I think) is a vmap that accepts an extra argument (mini_batch_size) and that starts looping instead of vectorizing, if the input size exceeds mini_batch_size. The problem is that now we are using vmap everywhere...

I don't think it's that easy (writing such a vmap would be relatively easy).
The problem is that you must make sure to loop over the minibatches only at the outermost level, and never call vmap on it.